### DISCUSSION TITLE
hysteria 2 能和 sslh 一起使用吗？ 例如443端口既提供SSH 远程登录，又提供https 伪装，还能 提供流量代理服务

### DISCUSSION THREAD

**Discussion by dengcong6666** [apernet/hysteria#1279] [Ideas]
hysteria 2 能和 sslh 一起使用吗？ 例如443端口既提供SSH 远程登录，又提供https 伪装，还能 提供流量代理服务

**Comment by dengcong6666** [comment]
官方说 启用混淆将使服务器与标准的 QUIC 连接不兼容，失去 HTTP/3 伪装的能力，我不太清楚如果打算配合 sslh使用是不是 会使代理失效

**Comment by haruue** [comment]
先把 TCP 和 UDP 分清楚再来参与讨论吧。

**Reply by dengcong6666** [reply]
> 先把 TCP 和 UDP 分清楚再来参与讨论吧。

我只是想问一下 sshd 和 hysteria 共用一个端口，有没有这种可能性？

**Reply by haruue** [reply]
TCP 443 （由 TCP 程序使用， 比如 https 服务器、 sshd 等）和 UDP 443 （由 UDP 程序使用， 比如 hysteria）本来就是两个不同的端口， 它们就不是「同一个端口」， 只是编号一样而已。

**Reply by dengcong6666** [reply]
Nginx 1.15.2版本开始，$ssl_preread_protocol变量被添加。该变量可以在[stream模块]中使用，用于在SSL/TLS握手阶段预先读取ClientHello消息，从而判断连接是SSL还是非SSL协议。通过这种方式，Nginx可以在同一个端口上转发不同的业务，例如同时处理[SSH]和[HTTPS]服务。

**Reply by dengcong6666** [reply]
> TCP 443 （由 TCP 程序使用， 比如 https 服务器、 sshd 等）和 UDP 443 （由 UDP 程序使用， 比如 hysteria）本来就是两个不同的端口， 它们就不是「同一个端口」， 只是编号一样而已。

我用nmap 扫描 hysteria 端口时，发现和正常的https 有些不同。 扫描 hysteria服务 的443端口，结果是 ssl/https? 而正常的https 服务不会多出这个？
![image](https://github.com/user-attachments/assets/cc60be10-d523-4c07-9a31-22f9db24fcc5)
![image](https://github.com/user-attachments/assets/75284d62-4564-451c-8856-6eae5a48cf3d)

**Reply by haruue** [reply]
Hysteria 只有在启用了 acme tlsAlpn 或者 masquerade listenHTTPS :443 的情况下才会监听 TCP 的 443 端口， 并且任何代理数据都不会通过 TCP 的 443 端口传输， 而是通过 UDP 的 443 端口传输。

你发的两张图中扫描的都是 TCP 443 端口而不是 UDP 443 端口， 你可以在你的服务器上执行 `ss -ltpn | grep 443` 来查询监听 TCP 443 端口的程序。

**Reply by dengcong6666** [reply]
> Hysteria 只有在启用了 acme tlsAlpn 或者 masquerade listenHTTPS :443 的情况下才会监听 TCP 的 443 端口， 并且任何代理数据都不会通过 TCP 的 443 端口传输， 而是通过 UDP 的 443 端口传输。
> 
> 你发的两张图中扫描的都是 TCP 443 端口而不是 UDP 443 端口， 你可以在你的服务器上执行 `ss -ltpn | grep 443` 来查询监听 TCP 443 端口的程序。

经过一天的折腾，已经可以了。思路是用 nginx 监听443 udp 端口，直接转发到 hysteria 容器 8443/udp 端口。监听443 tcp 端口，根据 $ssl_preread_protocol 值将流量  map 到 本地22 端口或者是  hysteria 容器 masquerade 端口  8443 (tcp)
